### ISSUE TITLE
[2.0] Discuss usage of `OASFilter` and `OASModelReader`

### ISSUE THREAD

**Issue by jmini** [eclipse/microprofile-open-api#300]
Discussion started by @arthurdm in https://github.com/eclipse/microprofile-open-api/issues/251#issuecomment-442157779 :

---

I would like to throw in the idea of removing `OASFilter` and `OASModeReader` from the spec in the `2.0` version.   The benefits would be an easier-to-understand spec, and would allow full compatibility with an `offline` (via Maven or other) generation of the OAS3 doc.  

I believe that modern usage of OAS3 is either via top-down / design-first, where you have the doc generated by a design team, or bottom-up / code-first, where the developer iteratively works on the JAX-RS app, generating it every time. 

In both cases there is a point in time where development is frozen and no more changes to the API should be allowed.   Therefore, why would we want to use a dynamic filter or model builder?     The only dynamic setting I can reasonably accept are changes to the `server` element, which can already be achieved by configuring the `mp.openapi.servers.*` related keys via MP Config.

**Comment by jmini** [comment]
The use case to be able to compute the OpenAPI during the build seems reasonable.

---
This was discussed here:
https://github.com/eclipse/microprofile-open-api/issues/294

---

I had a quick look into the http://micronaut.io/ approach and they are computing the OpenAPI spec at compile time.

---

I did not investigate the `OASFilter` feature yet (because we are still using `swagger-jaxrs` (or something similar) from [swagger-core](https://github.com/swagger-api/swagger-core)) but here is the feature I had in mind:

For each endpoints, we know if you can do them (pseudo-description):

```
GET /booking/{id}
    // Read booking entry of a given ID
    // Can be performed by REGULAR_USER and ADMIN

DELETE /booking/{id}
    // Delete booking entry with ID
    // Can be performed only by ADMIN

```

In the real OpenAPI specification for this, we have introduced a custom extension to our operations: `x-required-role`. Example for create:

```yaml
  /booking/create:
    post:
      operationId: bookingsCreate
      requestBody:
        description: The entity to create
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/Booking'
        required: true
      responses:
        200:
          description: The entity as it exists now in the system
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Booking'
        default:
          $ref: '#/components/responses/ErrorResponse'
      x-required-role: ADMIN
```

I was hoping that I can leverage the `OASFilter` feature in order to serve an OpenAPI specification that depends from the logged-in user. If an `ADMIN` watch the OpenAPI Spec, he see all the operations. If a  `REGULAR_USER` loads the OpenAPI Spec then he see only the operation he can performs.

When nobody is logged in, he sees the complete spec (the same that you would compute during compile time).

It this a valid use case for `OASFilter`? Is this misusing the feature?

**Comment by arthurdm** [comment]
hey @jmini - I think the scenario of displaying operations based on the logged in user is a very cool idea.   Instead of doing this manually via a filter though, I would propose we investigate how to augment our existing annotations to allow for such scenario.  For example, perhaps this is something we add into `SecurityRequirement`?

**Comment by charleech** [comment]
I apologize for interrupting you. 

I'm worried about the wording "remove" and would like to share my `OASFilter` use case for your consideration.

At the moment I'm using the `OASFilter` with `system properties` or `environment variable` to filter the URL for `@SecurityScheme`, e.g. the `authorizationUrl`, `tokenUrl`, `refreshUrl` , `OpenIdConnectUrl`. Because in my environment, dev, sit, uat, prod may use, different host/url, especially a different host where my services are deployed and not in my control. The ability to change those values help me a lot.

The example for my `OASFilter` is as the following: -

```java
public class DemoOASFilter implements OASFilter {
    @Override
    public SecurityScheme filterSecurityScheme(final SecurityScheme securityScheme) {
        Config config = ConfigProvider.getConfig();
        String myUrl  = config.getValue("my.url", 
                                        String.class);
        securityScheme.getFlows().
                       getImplicit().
                       setAuthorizationUrl(myUrl);

        return securityScheme;
    }
}
```

The example while deploying is as the following: -

```bash
docker run -it \
    --env my.url=http://some/url \
    ....

java -Dmy.url=http://some/url -jar  ....
```

**Comment by arthurdm** [comment]
thanks for the feedback @charleech!  That's a good example of the filter usage.

**Comment by arthurdm** [comment]
I have renamed the title, so it's more of a discussion to gather use cases =)

**Comment by arthurdm** [comment]
I found out that MicroProfile Fault Tolerance uses a [very cool](https://download.eclipse.org/microprofile/microprofile-fault-tolerance-2.0/microprofile-fault-tolerance-spec.html#_config_fault_tolerance_parameters) pattern for overriding any annotation value.  

Basically  it is setting `<classname>/<methodname>/<annotation>/<parameter>`

For example: `com.acme.test.MyClient/serviceA/Retry/maxDuration=3000`

So I wonder if we can use this pattern for the scenario that @charleech provided, and for most (if not all) cases where a Filter would be needed?    Thoughts - @jmini / @EricWittmann

**Comment by jmini** [comment]
@arthurdm: I am not sure how yet, but the MP-OpenAPI spec could plan a way that gets the OpenAPI Spec at compile time (I think it was your use case and something that framework like micronaut are also doing) with the limitation that the output at runtime might be modified by `OASFilter` and `OASModeReader`. 

Maybe a config to enable runtime modifications (could be false by default).

This way, vendors that would like to compute OpenAPI spec at compile could do it.

**Comment by arthurdm** [comment]
that's right @jmini - it would be very nice to have a choice of generating the OAS3 doc at compile time or at runtime.  

 Perhaps we can look at in terms of compile capabilities vs runtime capabilities:

**Compile**
- supports static OAS3 doc
- support annotations
- support MP config (this is where we can allow for generic annotation overrides, using the same concept as MP Fault Tolerate)

**Runtime**
- everything from compile
- the `/openapi` endpoint (and future extensions, like the SPIs coming in `2.0`)
- `OASFilter` and `OASModelReader`

**Comment by EricWittmann** [comment]
I'm in favor of this approach in principal.  But is there really any reason why `OASFilter` and `OASModelReader` can't be supported at compile time?  I mean obviously it will depend greatly on the provided class, but I can imagine use-cases where these interfaces could be useful without requiring a running system.

**Comment by charleech** [comment]
I would like to inform us the further use case for `java.util.Map` by following the [Dictionaries, HashMaps and Associative Arrays](https://swagger.io/docs/specification/data-models/dictionaries/) as the following: -

```java
public class DemoOASFilter implements OASFilter {

    @Override
    public void filterOpenAPI(final OpenAPI openAPI) {
        openAPI.getComponents().
                getSchemas().
                put("SimpleMap",
                    this.createMapKey());
    }
    
    private Schema createMapKey() {
        return OASFactory.createObject(Schema.class).
                          type(SchemaType.OBJECT).
                          additionalProperties(
                              OASFactory.createObject(Schema.class).
                                         type(SchemaType.STRING)
                          );         
    }
}
```

```java
@Path("/echo")
public class EchoResource {
    @POST
    @Consumes({
        MediaType.APPLICATION_JSON
    })
    @Produces({
        MediaType.APPLICATION_JSON
    })
    @Operation(
        operationId = "echoPost",
        summary     = "Echo with HTTP POST.",
        description = "Echo with HTTP POST"
    )
    @APIResponses(
        value = {
            @APIResponse(
                responseCode = "200",
                description  = "The echo data",
                content      = {
                    @Content(
                        mediaType = "application/json",
                        schema    = @Schema(
                            ref = "SimpleMap"
                        )
                    )
                }
            )
        }
    )
    public Response echoPost(@RequestBody(
                                 description = "The posting data",
                                 content = {
                                     @Content(
                                         mediaType = MediaType.APPLICATION_JSON,
                                         schema = @Schema(
                                             ref = "SimpleMap"
                                         )
                                     )
                                 }
                             ) 
                             Map<String, String> input) {
        
        return Response.ok(input).build();
    }
```

```json
{
    "openapi": "3.0.0",
    "info": {
        "title": "Demo App",
        "description": "A Demo Application.",
        "version": "1.0"
    },
    "paths": {
      "/api/echo": {
        "post" : {
                "summary": "Echo with HTTP POST.",
                "description": "Echo with HTTP POST.",
                "operationId": "echoPost",
                "requestBody": {
                    "description": "The posting data",
                    "content": {
                        "application/json": {
                            "schema": {
                                "$ref": "#/components/schemas/SimpleMap"
                            }
                        }
                    }
                },
                "responses": {
                    "200": {
                        "description": "The echo data",
                        "content": {
                            "application/json": {
                                "schema": {
                                    "$ref": "#/components/schemas/SimpleMap"
                                }
                            }
                        }
                    }
                }
        }
      }
    },
    "components": {
        "schemas": {
            "SimpleMap": {
                "type": "object",
                "additionalProperties": {
                    "type": "string"
                 }
            }
        }
    }
}
```

**Comment by jfbenckhuijsen** [comment]
Adding to the discussion on the filtering of the API based on the logged in user, this is actually a use case I'm facing in my clients project. So +1 from my side. 

With regards to the suggestion of not-logged-in == full API, is this something which can maybe be configurable? From a security point of view, not being logged in on a private API (which in a b2b scenario is often the case), would mean you're not entitled to ANY API call, so I would expect not seeing anything in that case. Of course, this can be handled at e.g. an API Gateway level, but having an  extra level of security might be helpful.

**Comment by jloupdef** [comment]
Hi, were there any changes on this subject, I am currently experimenting with quarkus and tried to use OAS filter to change auth url from environment variable and could not get it to work at runtime, it seams to be evaluated at built time even if you are discussing here that it is not the case and cannot be...
The code is pretty simple :
```java
public class OASSecurityConfiguration implements OASFilter {
    @Override
    public SecurityScheme filterSecurityScheme(final SecurityScheme securityScheme) {
        securityScheme.getFlows().
                       getImplicit().
                       setAuthorizationUrl(ConfigProvider.getConfig().getValue("quarkus.oidc.auth-server-url", String.class)+"/protocol/openid-connect/auth");
        return securityScheme;
    }
}
```

**Comment by EricWittmann** [comment]
Unless Quarkus works differently than my understanding (possible) then this filter should be executed at runtime, although only once - either on startup or possibly the first time the `/openapi` endpoint is hit.  I would be very surprised if this is being executed at **build** time.

What happens when you run this code?

**Comment by jloupdef** [comment]
If I put a log in the function, the log can be seen during build and is never displayed at runtime, that is why I am sure of my analysis, as you can see :
![image](https://user-images.githubusercontent.com/22762/73396551-ed07af00-42e1-11ea-9ade-b08735bd01ed.png)

**Comment by EricWittmann** [comment]
Well, in that case I was definitely wrong about how Quarkus works with SmallRye OpenAPI (which is the implementation of this spec that Quarkus uses).  So I guess the next suggestion is to raise this as a question/issue in Quarkus.  It might be useful/necessary to have a Quarkus-specific feature allowing runtime modification of the OpenAPI doc.

**Comment by radcortez** [comment]
Quarkus applies it own extension on top of SmallRye OpenAPI to resolve everything at build time. Maybe this is a case that it was not considered, about setting specific Config in the filter that needs to be evaluated at runtime.

**Comment by msmiths** [comment]
> Quarkus applies it own extension on top of SmallRye OpenAPI to resolve everything at build time. Maybe this is a case that it was not considered, about setting specific Config in the filter that needs to be evaluated at runtime.

@radcortez Is this still the way that Quarkus works with respect to SmallRye OpenAPI?  I believe that the intention of both the `OASModelReader` and `OASFilter` interfaces is that they can be used by applications to dynamically build/filter the OpenAPI model at runtime, based on information that is available in the runtime environment.  The various examples outlined in this issue require that the implementations of these interfaces are invoked at runtime in order for them to work correctly.  Is the Quarkus extension being updated to allow this?  Has this change already been made?

**Comment by lscoughlin** [comment]
I have a number of use cases where parts of the api are generate in full or in part based on user configuration of the system at runtime.  Removing OASModelReader in particular would pose an enormous inconvenience to me.

Runtime mutability is pretty much exactly what these classes were included for.  You're not required to do anything with them if you don't understand them but please do not propose removing things just because you personally don't have a use case.

**Comment by MikeEdgar** [comment]
This issue was reviewed in today's discussion and based on this discussion (including user feedback/comments) the consensus was that the current functionality of the `OASFilter` and `OASModelReader` should remain as-is. Please reopen this issue if you disagree or would like to discuss further.
